Audit logging

woodruffw opened the pull request:
Adds `UserEvent` and `ProjectEvent` models, as well as corresponding helpers for recording events.

A shortlist of loggable events (not all of which need to be part of this PR):

Users: 

- [x] Login (+ whether a 2FA method was used)
- [x] Token creation/deletion
- [x] 2FA method creation/deletion
- [x] Email addition/removal/verification/deverification
- [x] Password change

Projects:

- [x] Version upload/removal
- [x] Change in owner/maintainers
- [x] Token creation by a user with the project in scope

TODO:

- [x] Add `assert`s to `record_event` calls in unit tests, where appropriate.
- [x] Restrict retrieval of user events to a recent timespan (last two weeks?)
- [x] Allow Warehouse admins to view the events for a user/project


Summary of events added:

Users:
* `account:email:add` (New email added to account)
* `account:email:remove` (Email removed from account)
* `account:email:primary:change` (Primary email changed or added)
* `account:email:reverify` (Email reverified)
* `account:password:change` (Account password changed)
* `account:two_factor:method_added` (Two factor method added)
* `account:two_factor:method_removed` (Two factor method removed)
* `account:api_token:added` (API token added)
* `account:api_token:removed` (API token removed)
* `account:create` (Account created)
* `account:password:reset:request` (Password reset request sent)
* `account:password:reset` (Password reset request completed)
* `account:email:verified` (Email address verified)
* `account:login:success` (Successful login)


Projects:
* `project:create` (Project created)
* `project:release:added` (New project release added)
* `project:release:removed` (Project release removed)
* `project:release:file:removed` (Project release file removed)
* `project:api_token:added` (API token added w/ project in scope)
* `project:api_token:removed` (API token removed w/ project in scope)
* `project:role:add` (New role added to project)
* `project:role:delete` (Role removed from project)
* `project:role:change` (Role changed in project)

cc @brainwane @ewdurbin @di @dstufft @nlhkabu 

Closes #5863.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -237,6 +247,7 @@ def test_post_validate_no_redirects(
             find_userid=pretend.call_recorder(lambda username: 1),
             update_user=lambda *a, **k: None,
             has_two_factor=lambda userid: False,
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -275,6 +287,7 @@ def test_two_factor_auth(self, pyramid_request, redirect_url, token_service):
             find_userid=pretend.call_recorder(lambda username: 1),
             update_user=lambda *a, **k: None,
             has_two_factor=lambda userid: True,
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -406,6 +420,7 @@ def test_totp_auth(self, monkeypatch, pyramid_request, redirect_url):
             has_totp=lambda userid: True,
             has_webauthn=lambda userid: False,
             check_totp_value=lambda userid, totp_value: True,
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this call_recorder.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -468,6 +484,7 @@ def test_totp_auth_invalid(self, pyramid_request, redirect_url):
             has_totp=lambda userid: True,
             has_webauthn=lambda userid: False,
             check_totp_value=lambda userid, totp_value: False,
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this call_recorder.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -870,9 +890,11 @@ def test_register_redirect(self, db_request, monkeypatch):
                 create_user=create_user,
                 add_email=add_email,
                 check_password=lambda pw, tags=None: False,
+                record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this call_recorder.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -951,10 +973,12 @@ def test_request_password_reset(
+        pyramid_request.remote_addr = "0.0.0.0"
         token_service.dumps = pretend.call_recorder(lambda a: "TOK")
         user_service.get_user_by_username = pretend.call_recorder(lambda a: stub_user)
+        user_service.record_event = pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this call_recorder.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -997,12 +1021,16 @@ def test_request_password_reset_with_email(
         token_service.dumps = pretend.call_recorder(lambda a: "TOK")
         user_service.get_user_by_username = pretend.call_recorder(lambda a: None)
         user_service.get_user_by_email = pretend.call_recorder(lambda a: stub_user)
+        user_service.record_event = pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/accounts/test_views.py:
```diff
@@ -1046,16 +1074,19 @@ def test_request_password_reset_with_non_primary_email(
         token_service.dumps = pretend.call_recorder(lambda a: "TOK")
         user_service.get_user_by_username = pretend.call_recorder(lambda a: None)
         user_service.get_user_by_email = pretend.call_recorder(lambda a: stub_user)
+        user_service.record_event = pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -186,7 +194,8 @@ def test_add_email(self, monkeypatch, pyramid_config):
         user_service = pretend.stub(
-            add_email=pretend.call_recorder(lambda *a, **kw: email)
+            add_email=pretend.call_recorder(lambda *a, **kw: email),
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -262,6 +272,9 @@ def test_add_email_validation_fails(self, monkeypatch):
         email = pretend.stub(id=pretend.stub(), primary=False, email=pretend.stub())
         some_other_email = pretend.stub()
+        user_service = pretend.stub(
+            record_event=pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -341,13 +355,17 @@ def test_delete_email_is_primary(self, monkeypatch):
 
-        db_request.find_service = lambda *a, **kw: pretend.stub()
+        user_service = pretend.stub(
+            record_event=pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -374,8 +392,12 @@ def test_change_primary_email_without_current(self, monkeypatch, db_request):
 
-        db_request.find_service = lambda *a, **kw: pretend.stub()
+        user_service = pretend.stub(
+            record_event=pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -414,7 +436,13 @@ def test_change_primary_email_not_found(self, monkeypatch, db_request):
+            verified=False,
+            email="email_address",
+            user=pretend.stub(
+                record_event=pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -499,7 +528,8 @@ def test_change_password(self, monkeypatch):
         user_service = pretend.stub(
-            update_user=pretend.call_recorder(lambda *a, **kw: None)
+            update_user=pretend.call_recorder(lambda *a, **kw: None),
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -847,6 +878,7 @@ def test_validate_totp_provision(self, monkeypatch):
         user_service = pretend.stub(
             get_totp_secret=lambda id: None,
             update_user=pretend.call_recorder(lambda *a, **kw: None),
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -990,6 +1023,7 @@ def test_delete_totp(self, monkeypatch, db_request):
         user_service = pretend.stub(
             get_totp_secret=lambda id: b"secret",
             update_user=pretend.call_recorder(lambda *a, **kw: None),
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -1157,7 +1192,8 @@ def test_get_webauthn_options(self):
         user_service = pretend.stub(
-            add_webauthn=pretend.call_recorder(lambda *a, **kw: pretend.stub())
+            add_webauthn=pretend.call_recorder(lambda *a, **kw: pretend.stub()),
+            record_event=pretend.call_recorder(lambda *a, **kw: None),
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -1242,6 +1279,9 @@ def test_validate_webauthn_provision_invalid_form(self, monkeypatch):
 
     def test_delete_webauthn(self, monkeypatch):
+        user_service = pretend.stub(
+            record_event=pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented on tests/unit/manage/test_views.py:
```diff
@@ -1583,16 +1743,21 @@ def test_delete_macaroon(self, monkeypatch):
             ),
         )
+        user_service = pretend.stub(
+            record_event=pretend.call_recorder(lambda *a, **kw: None)
```
TODO(ww): Add an assert for this `call_recorder`.

woodruffw commented:
@nlhkabu this should be ready for your input on UI/X 😄

woodruffw commented:
NB: We currently record `project:api_token:added` and `project:api_token:removed` whenever a token with the given project in-scope is added or removed, but we don't do the same when a user-scoped token is added or removed (we only record that for the user). Adding that to the project would constitute a minor info-leak (other project owners would be able to see token creation/deletion events for potentially unrelated tokens created by other owners), but that might be an acceptable tradeoff.

brainwane commented:
@woodruffw I see this PR does _not_ include the upload or removal of a specific release/file, per the checkboxes on the initial comment; should we expect that you're going to add that in this PR, or that implementing that as an auditable event would wait for a future PR?

woodruffw commented:
> should we expect that you're going to add that in this PR, or that implementing that as an auditable event would wait for a future PR?

My plan was to put it in a future PR; it involves a different set of routes (`forklift/legacy.py`) and, while basically identical to this changeset, would balloon it a bit more. Is that alright?

brainwane commented:
Yes, I think that's fine! I just wanted to check. :)

woodruffw commented on warehouse/templates/manage/account.html:
```diff
@@ -484,25 +484,113 @@ <h2>API tokens <span class="badge badge--warning">Beta feature</span></h2>
     {% if recent_events|length > 0 %}
-    <table>
+
+    {% macro event_summary(event) -%}
```
NB: I think we could probably come up with a Jinja2 filter here to perform event prettification for us; I can take a look at that in a bit.

nlhkabu replied:
that would be great :grin:

nlhkabu commented:
UI updates:

![Screenshot from 2019-08-13 21-23-21](https://user-images.githubusercontent.com/3323703/62974683-ac3ddb80-be10-11e9-9c75-8070b899abf3.png)

![Screenshot from 2019-08-13 21-22-58](https://user-images.githubusercontent.com/3323703/62974666-9f20ec80-be10-11e9-9847-b99978050095.png)

brainwane commented:
@nlhkabu I assume that when you hover over the relative timestamp you get an absolute time?

nlhkabu commented:
Nope, not currently @brainwane - I'll add this when I fix all other time/date instances in https://github.com/pypa/warehouse/issues/6312

ewdurbin commented on warehouse/accounts/models.py:
```diff
@@ -95,6 +96,18 @@ class User(SitemapMixin, db.Model):
     )
 
+    events = orm.relationship(
+        "UserEvent", backref="user", cascade="all, delete-orphan", lazy=False
```
I think this is _probably_ the right relationship, but I'm not sure about the cascade. It seems like we might want to keep these for some period post deletion? For instance in the event of an attacker taking control of an account and purging all projects then deleting the account.

woodruffw replied:
Good point! It looks like the default `cascade` is `"save-update, merge"`, so it looks like that's what we want IIUC.

ewdurbin commented on warehouse/accounts/interfaces.py:
```diff
@@ -179,6 +179,13 @@ def get_webauthn_by_credential_id(user_id, credential_id):
         or None of the user doesn't have a credential with this ID.
         """
 
+    def record_event(user_id, **kwargs):
```
Should this interface be more strict than `**kwargs`? I realize it's more or less passing through whatever comes its way to the UserEvent object itself, though a better docstring and explicit signature would be beneficial at this definition.

woodruffw replied:
Yep. I made that interface while still hammering the model out; I'll make the keywords explicit and add the appropriate documentation.

ewdurbin commented on warehouse/manage/views.py:
```diff
@@ -593,12 +656,42 @@ def create_macaroon(self):
+                    "caveats": macaroon_caveats,
+                },
+            )
+            if "projects" in form.validated_scope:
```
I'm wondering if we also want to expose when a user creates an API token that has the same effective permissions, in other words a "user scoped" token. It seems like it might not be the _best_ idea, but it's also probably not relevant to this PR at this time.

woodruffw replied:
Yep, that was the topic of this NB: https://github.com/pypa/warehouse/pull/6339#issuecomment-520879774

One minor issue with recording project events on a "user scoped" token's creation is that only preexisting projects will receive the creation event, while future projects will silently allow the token but not contain the corresponding event. That's not necessarily worse than the current PR behavior, just something to keep in mind. We could also generally mitigate the problem by listing associated API tokens separately somewhere in the project management view.

nlhkabu replied:
In the interest of getting this PR merged, I've opened a new ticket to follow up on this: https://github.com/pypa/warehouse/issues/6441

ewdurbin commented on warehouse/packaging/models.py:
```diff
@@ -135,6 +136,10 @@ class Project(SitemapMixin, db.Model):
     )
 
+    events = orm.relationship(
+        "ProjectEvent", backref="project", cascade="all, delete-orphan", lazy=False
```
Similar concern as the User/UserEvent relationship. We likely want to maintain the log post project deletion for the retention period.

ewdurbin commented on warehouse/manage/views.py:
```diff
@@ -1073,24 +1222,24 @@ def manage_project_history(project, request):
-        .options(joinedload("submitted_by"))
-        .filter(JournalEntry.name == project.name)
-        .order_by(JournalEntry.submitted_date.desc(), JournalEntry.id.desc())
+    events_query = (
```
I think we still want to display the JournalEntry data in this view, until we decide on how we'll handle either migrating this data into ProjectEvent objects, or otherwise deprecating it.

woodruffw replied:
Hmm, displaying both the `JournalEntry` and `ProjectEvent` tables in the same view might be a bit difficult because of the pagination generator. Would a separate tab like "Legacy Journal" (or similar) be alright?

brainwane replied:
A separate tab is fine with me.

ewdurbin replied:
Certainly not in the same table, no. A separate Tab seems best and cleanest until we decide the fate of Journal

woodruffw replied:
Done!

ewdurbin commented on warehouse/templates/manage/history.html:
```diff
@@ -17,38 +17,64 @@
     </thead>
     <tbody>
-      {% for journal in journals %}
+      {% for event in events %}
```
I think we want to keep Journals visible, and display them in a separate table. Once more, to expose this information until we either migrate JournalEntry objects into ProjectEvent objects or otherwise deprecate them.

nlhkabu replied:
Thanks @ewdurbin.  FYI we have a ticket discussing this here : https://github.com/pypa/warehouse/issues/6425

ewdurbin replied:
Just found that and commented. 😬

woodruffw commented:
@nlhkabu I've added the remaining project events for release changes (as well as project creation itself):

* `project:create`
* `project:release:add`
* `project:release:remove`
* `project:release:file:remove`

nlhkabu commented:
Thanks @woodruffw 

I've formatted the new events in the security history page - but I wasn't sure how I could test these...  Would you be able to generate a screenshot on your side?

Also, I was wondering - is there a technical reason why we don't capture/expose data re: the user who has created the release, removed the release and removed the file?  I would have thought this is pretty useful information to expose.

woodruffw commented:
> but I wasn't sure how I could test these... Would you be able to generate a screenshot on your side?

For `project:create` and `project:release:add`, you should be able to test them by creating a token and uploading a new project to your local instance. For `project:release:remove` and `project:release:file:remove`, you should be able to use the management view to delete an extant release or particular file, which should trigger the event. I'll also take a screenshot in a bit 🙂 



> Also, I was wondering - is there a technical reason why we don't capture/expose data re: the user who has created the release, removed the release and removed the file? I would have thought this is pretty useful information to expose.

Nope, just an oversight on my part. I'll expose this information.

woodruffw commented:
Here's what the new events look like:

<img width="851" alt="Screen Shot 2019-08-15 at 10 53 20 AM" src="https://user-images.githubusercontent.com/3059210/63103347-edd2a180-bf4a-11e9-8ae1-87074f36464e.png">

di commented on warehouse/templates/manage/history.html:
```diff
@@ -17,38 +17,89 @@
+    </small>
+    {% elif event.tag == "project:release:file:remove" %}
+    <strong>File removed from release version {{ event.additional.canonical_version }}</strong><br>
+    <small>Filename: {{ event.additional.filename }}</small><br>
```
```suggestion
    <small>Filename: <code>{{ event.additional.filename }}</code></small><br>
```

woodruffw commented:
Woot!
